validate ShiftEvent#occurred_at via Shift#day

bjmllr opened the pull request:
Fixes #68

bjmllr commented:
Well, the validation seems to work, but it also makes the build fail. I can't reproduce the failure on travis locally, any ideas?

leesharma commented:
I'd guess it has to do with line 4 of the shift_event factory:

```ruby
  occurred_at { Faker::Time.between(DateTime.now - 1, DateTime.now) }
```

I fixed that in my local branch by changing it to:
```ruby
  occurred_at { Faker::Time.between(day.beginning_of_day, day.end_of_day) }
```

Not sure why it's not failing locally. Did you restart guard/spring before running specs?

leesharma commented:
Also, this is a preference, but since the validator is pretty specific to this case/class, I'd personally do it as a method instead of a full class. There's a lot less code to sift through then.

```ruby
class ShiftEvent < ActiveRecord::Base
  validate :shift_day_matches_occurred_at

  # ...

  def shift_day_matches_occurred_at
    return if day == occurred_at.to_date
    msg = "occurred_at (#{occurred_at.to_date}) does not match day (#{day})"
    errors.add(:day, msg)
  end
end
```

bjmllr commented:
Replaced validation class with validation method.

Fixed all the `Faker::Time.between` calls to use times rather than dates. We may still have random failures if there is a time zone issue hiding here.

leesharma commented on app/models/shift_event.rb:
```diff
@@ -1,4 +1,6 @@
 class ShiftEvent < ActiveRecord::Base
+  include ActiveModel::Validations
```
I can't test this at the moment, but I'm pretty sure we don't need this line.

leesharma commented on spec/factories/shift.rb:
```diff
@@ -6,8 +6,11 @@
+
         ShiftEvent::ACTIONS.keys.zip(
-          Array.new(4) { Faker::Time.between(shift.day, shift.day) }.sort
+          Array.new(4) { Faker::Time.between(day_begin, day_end) }.sort
```
Since it looks like we're using the same logic here as is in the `:shift_event` factory, can we just leave setting the `occurred_at` out altogether? Or are we relying on the ordering?

bjmllr replied:
We `zip` with `ACTIONS` to get one event of each action-type, so the `sort` ensures that the sign-in occurs before the sign-out.

bjmllr commented:
Thanks for reviewing!
